### ISSUE TITLE
Idea: Speed up scheduler by removing socket poll call and reduce amount of routine calls

### ISSUE THREAD

**Issue by darenliang** [yzard/scaled#1]
I was doing some profiling and found out that it might not be suitable to use asyncio.gather as certain routines are no-ops in an overwhelming majority of the time. It might be better to run loops within the event loop to have more control over when and where a routine is called.

This would also eliminate the need for polling on the socket as recv_multipart already forces a yield back to the event loop.

```py
async def routine(self):
    count = await self._socket.poll(POLLING_TIME_MILLISECONDS)
    if not count:
        return

    for _ in range(count):
        frames = await self._socket.recv_multipart()
        if not self.__is_valid_message(frames):
            continue

        source, message_type_bytes, payload = frames[0], frames[1], frames[2:]
        message_type = MessageType(message_type_bytes)
        self.__count_one("received", message_type)
        message = PROTOCOL[message_type_bytes].deserialize(payload)
        await self._callback(source, message_type, message)
```

Can turn into:

```py
async def routine(self):
    frames = await self._socket.recv_multipart()
    if not self.__is_valid_message(frames):
        continue
    
    source, message_type_bytes, payload = frames[0], frames[1], frames[2:]
    message_type = MessageType(message_type_bytes)
    self.__count_one("received", message_type)
    message = PROTOCOL[message_type_bytes].deserialize(payload)
    await self._callback(source, message_type, message)
```

The general idea is to run a message loop that will receive and process messages and a cleanup loop that will perform any necessary clean up tasks. This allows certain routines to run less often, freeing the event loop to perform more performance critical routines.

I think task_manager's routine only needs to be called when a message is received so it would make sense to call it after the binder routine.

```py
def setup_loop(self, loop):

    async def on_message_loop():
        while True:
            await self._binder.routine()
            await self._task_manager.routine()
            await asyncio.sleep(0) # sleep(0) is optimized: https://docs.python.org/3/library/asyncio-task.html#asyncio.sleep

    async def cleanup_loop():
        while True:
            await self._function_manager.routine()
            await self._worker_manager.routine()
            await asyncio.sleep(1)  # its ok to have these routines run at longer intervals

    loop.create_task(on_message_loop())
    loop.create_task(cleanup_loop())

loop = get_event_loop(args.event_loop)
scheduler.setup_loop(loop)
loop.run_forever()
```

I didn't add in the signal event because it might be possible to avoid it by directly closing the loop in the signal callback:

```py
signal.signal(signal.SIGINT, lambda *args: loop.close())
```

We can see that a scheduler implementing the above changes can potentially be 30% faster.

Current scheduler running 100k no-op tasks
```
[INFO]2023-02-25 20:48:13-0500: logging to ('/dev/stdout',)
[INFO]2023-02-25 20:48:13-0500: ScaledClient: connect to tcp://127.0.0.1:2345
[INFO]2023-02-25 20:48:13-0500: beginning scaled submit 100000 tasks
[INFO]2023-02-25 20:48:16-0500: completed scaled submit 100000 tasks in 0:00:02.972396
[INFO]2023-02-25 20:48:16-0500: beginning scaled gather 100000 results
[INFO]2023-02-25 20:48:48-0500: completed scaled gather 100000 results in 0:00:32.169261
[INFO]2023-02-25 20:48:48-0500: ScaledClient: disconnect from tcp://127.0.0.1:2345
```

Improved scheduler running 100k no-op tasks
```
[INFO]2023-02-25 21:20:26-0500: logging to ('/dev/stdout',)
[INFO]2023-02-25 21:20:26-0500: ScaledClient: connect to tcp://127.0.0.1:2345
[INFO]2023-02-25 21:20:26-0500: beginning scaled submit 100000 tasks
[INFO]2023-02-25 21:20:28-0500: completed scaled submit 100000 tasks in 0:00:02.139044
[INFO]2023-02-25 21:20:28-0500: beginning scaled gather 100000 results
[INFO]2023-02-25 21:20:51-0500: completed scaled gather 100000 results in 0:00:22.793362
[INFO]2023-02-25 21:20:51-0500: ScaledClient: disconnect from tcp://127.0.0.1:2345
```

**Comment by yzard** [comment]
@darenliang, You are right, I am also doing research on nng, I think this thing is still not very mature, as user are far less, might be the most wanted feature that attract us is IPC and socket is thread safe

**Comment by darenliang** [comment]
Socket thread safety might not be as important as there is usually a thread safe implementation for zmq libraries. Also something you don't need to worry about in Python 😅.

**Comment by yzard** [comment]
The reason I do sending queue on connector is because of this Socket safety :), otherwise need maintain a lock

**Comment by yzard** [comment]
@darenliang , addressed and checkout the version==0.28
